include/circt/Dialect/FIRRTL/FIRRTLAnnotationHelper.h

```diff
@@ -246,6 +246,9 @@ struct ApplyState {
   InstancePathCache &instancePathCache;
   DenseMap<Attribute, FlatSymbolRefAttr> instPathToNLAMap;
   size_t numReusedHierPaths = 0;
+
+  llvm::StringMap<Value> legacyWiringSources;
+  llvm::StringMap<SmallVector<Value>> legacyWiringSinks;
```

Comment:
It would be slightly cleaner to define a new `struct LegacyWiringProblem` as opposed to doing the split of two string maps. Consider:

```c++
struct LegacyWiringProblem {
  source Value;
  sinks SmallVector<Value>;
};

DenseMap<LegacyWiringProblem> legacyWiringProblems;
```

Reply:
Also, there's usually not a huge benefit in using `llvm::StringMap` unless you have actual strings. Because you're dealing with `StringAttr`s (which are wrappers around interned strings), it's usually better to just use a `DenseMap<Attribute>` (and ignore the fact that this will allow for keys of any attribute type and not just `StringAttr`). This should be more efficient than `llvm::StringMap`.

Reply:
A logical issue I'm having with this is that the order of annotations impacts how we can structure data. If the sink for a pin is defined before a source for it, it's not possible to define a LegacyWiringProblem like this right away, we'd have to store something immediately and update it later. Would 
```
struct LegacyWiringProblem {
  Optional<Value> source;
  SmallVector<Value> sinks;
};
```
make any sense?

Reply:
`Optional<Value> source` is overkill as `Value` is nullable. I.e., it's totally fine to just leave `Value` uninitialized and it has boolean operations defined on it. Stuff like the following is totally fine:

```c++
LegacyWiringProblem foo = ...;
if (!foo.source)
  return mlir::emitError() << "no source set!";
```

Tangential tidbit: `llvm::Optional` is deprecated in favor of unifying around `std::optional`.